From e662fab3c1c042b4726cd3ced53f7fe54737ee17 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Wed, 31 Dec 2025 00:08:36 +0100
Subject: [PATCH 18/18] wintrust: Prevent checking if winex11/winewayland are
 signed.

Based on a patch from Lily, fixes DNA crashing at startup.
---
 dlls/wintrust/wintrust_main.c | 25 ++++++++++++++++++++++++-
 1 file changed, 24 insertions(+), 1 deletion(-)

diff --git a/dlls/wintrust/wintrust_main.c b/dlls/wintrust/wintrust_main.c
index a9b1587e5be..bb8a1e26f57 100644
--- a/dlls/wintrust/wintrust_main.c
+++ b/dlls/wintrust/wintrust_main.c
@@ -250,6 +250,23 @@ static DWORD WINTRUST_AddTrustStepsFromFunctions(struct wintrust_step *steps,
     return numSteps;
 }
 
+static BOOL needs_wintrust_fixes(void)
+{
+    static volatile char cache = -1;
+    if (cache == -1)
+    {
+        WCHAR path[MAX_PATH], *p, *name;
+        
+        GetModuleFileNameW(NULL, path, MAX_PATH);
+        name = path;
+        if ((p = wcsrchr(name, '/'))) name = p + 1;
+        if ((p = wcsrchr(name, '\\'))) name = p + 1;
+        
+        cache = !wcsicmp(name, L"EM-Win64-Shipping.exe");
+    }
+    return cache;
+}
+
 static LONG WINTRUST_DefaultVerify(HWND hwnd, GUID *actionID,
  WINTRUST_DATA *data)
 {
@@ -293,7 +310,13 @@ error:
     free(provData);
 
 done:
-    TRACE("returning %08lx\n", err);
+    if (needs_wintrust_fixes() && data->dwUnionChoice == 1 && 
+        (wcscmp(data->pFile->pcwszFilePath, L"C:\\windows\\system32\\winex11.drv") == 0 || 
+         wcscmp(data->pFile->pcwszFilePath, L"C:\\windows\\system32\\winewayland.drv") == 0))
+    {
+        TRACE(" checking for winex11/winewayland, skipping.. \n");
+        err = 0;
+    }
     return err;
 }
 
-- 
2.52.0

